### PR TITLE
Add `image_id` to CRI `Container` message

### LINKED ISSUE #7143: imageRef in container status is not image ID and lead to kubelet image gc work abnormally
### What happened?

The imageRef of containers, returned by runtime cri-o, may not be image ID. This will mislead kubelet's image-gc logical.

In [CRI spec](https://github.com/kubernetes/kubernetes/blob/90c362b3430bcbbf8f245fadbcd521dab39f1d7c/staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.pb.go#L5544), it consider imageRef as image ID:
```
// Reference to the image in use. For most runtimes, this should be an
// image ID.
ImageRef string `protobuf:"bytes,5,opt,name=image_ref,json=imageRef,proto3" json:"image_ref,omitempty"`
```

When do image gc logical, kubelet will get image and container status from cri-o runtime service. And then, compare currently used images (by container) and existing images by image ID.

If imageRef return by cri-o is not image ID, kubelet image gc can not work properly.

Kubelet's key logs may look like this:
```
I0719 16:36:14.498634 xxx image_gc_manager.go:233] "Container uses image" pod="multus/multus-admission-controller-7ff9c677b7-2p6lx" containerName="multus-admission-controller" containerImage="38da782fd6a202da62c652532b5d289278819d84bf2f7280ad40d214903db6e3" imageID="image.foo.bar/multus-admission-controller@sha256:36f44236d5c1aa57bd5e930179f020c089e2bde0d20a5d349f5af8fefe2d6235"
I0719 16:36:14.500180 xxx image_gc_manager.go:244] "Adding image ID to currentImages" imageID="38da782fd6a202da62c652532b5d289278819d84bf2f7280ad40d214903db6e3"
I0719 16:36:14.500186 xxx image_gc_manager.go:261] "Image ID has size" imageID="38da782fd6a202da62c652532b5d289278819d84bf2f7280ad40d214903db6e3" size=316580906
I0719 16:36:14.500192 xxx image_gc_manager.go:264] "Image ID is pinned" imageID="38da782fd6a202da62c652532b5d289278819d84bf2f7280ad40d214903db6e3" pinned=false
I0719 16:36:14.534022 xxx image_gc_manager.go:371] "Evaluating image ID for possible garbage collection" imageID="38da782fd6a202da62c652532b5d289278819d84bf2f7280ad40d214903db6e3"
I0719 16:36:14.534031 xxx image_gc_manager.go:387] "Removing image to free bytes" imageID="38da782fd6a202da62c652532b5d289278819d84bf2f7280ad40d214903db6e3" size=316580906
E0719 16:36:14.534609 xxx remote_image.go:270] "RemoveImage from image service failed" err="rpc error: code = Unknown desc = Image used by c5ea50c909f1794ec8cf7e10e3e740bbd76f4d27d8c8af7625f84ef844868128: image is in use by a container" image="38da782fd6a202da62c652532b5d289278819d84bf2f7280ad40d214903db6e3"
E0719 16:36:14.534633 xxx kuberuntime_image.go:122] "Failed to remove image" err="rpc error: code = Unknown desc = Image used by c5ea50c909f1794ec8cf7e10e3e740bbd76f4d27d8c8af7625f84ef844868128: image is in use by a container" image="38da782fd6a202da62c652532b5d289278819d84bf2f7280ad40d214903db6e3"
```

The real image ID is *38da782fd6a202da62c652532b5d289278819d84bf2f7280ad40d214903db6e3*, and image name is *image.foo.bar/multus-admission-controller@sha256:36f44236d5c1aa57bd5e930179f020c089e2bde0d20a5d349f5af8fefe2d6235*. However, kubelet just get the opposite way:
```
I0719 16:36:14.498634 xxx image_gc_manager.go:233] "Container uses image" ... containerImage="38da782fd6a202da62c652532b5d289278819d84bf2f7280ad40d214903db6e3" imageID="image.foo.bar/multus-admission-controller@sha256:36f44236d5c1aa57bd5e930179f020c089e2bde0d20a5d349f5af8fefe2d6235"
```

### What did you expect to happen?

imageRef returned by cri-o in container status should be image ID, and be consistent with image id information return by runtime image service.

### How can we reproduce it (as minimally and precisely as possible)?

Making a kubernetes node under disk pressure, and trigger kubelet image gc.

Image gc would continously try to remove image, which is used by container, and failed. The kubelet logs (log level 5) may continously output like this:
```
I0719 16:36:14.498634 xxx image_gc_manager.go:233] "Container uses image" pod="multus/multus-admission-controller-7ff9c677b7-2p6lx" containerName="multus-admission-controller" containerImage="38da782fd6a202da62c652532b5d289278819d84bf2f7280ad40d214903db6e3" imageID="image.foo.bar/multus-admission-controller@sha256:36f44236d5c1aa57bd5e930179f020c089e2bde0d20a5d349f5af8fefe2d6235"
I0719 16:36:14.500180 xxx image_gc_manager.go:244] "Adding image ID to currentImages" imageID="38da782fd6a202da62c652532b5d289278819d84bf2f7280ad40d214903db6e3"
I0719 16:36:14.500186 xxx image_gc_manager.go:261] "Image ID has size" imageID="38da782fd6a202da62c652532b5d289278819d84bf2f7280ad40d214903db6e3" size=316580906
I0719 16:36:14.500192 xxx image_gc_manager.go:264] "Image ID is pinned" imageID="38da782fd6a202da62c652532b5d289278819d84bf2f7280ad40d214903db6e3" pinned=false
I0719 16:36:14.534022 xxx image_gc_manager.go:371] "Evaluating image ID for possible garbage collection" imageID="38da782fd6a202da62c652532b5d289278819d84bf2f7280ad40d214903db6e3"
I0719 16:36:14.534031 xxx image_gc_manager.go:387] "Removing image to free bytes" imageID="38da782fd6a202da62c652532b5d289278819d84bf2f7280ad40d214903db6e3" size=316580906
E0719 16:36:14.534609 xxx remote_image.go:270] "RemoveImage from image service failed" err="rpc error: code = Unknown desc = Image used by c5ea50c909f1794ec8cf7e10e3e740bbd76f4d27d8c8af7625f84ef844868128: image is in use by a container" image="38da782fd6a202da62c652532b5d289278819d84bf2f7280ad40d214903db6e3"
E0719 16:36:14.534633 xxx kuberuntime_image.go:122] "Failed to remove image" err="rpc error: code = Unknown desc = Image used by c5ea50c909f1794ec8cf7e10e3e740bbd76f4d27d8c8af7625f84ef844868128: image is in use by a container" image="38da782fd6a202da62c652532b5d289278819d84bf2f7280ad40d214903db6e3"
```

### Anything else we need to know?

_No response_

### CRI-O and Kubernetes version

<details>

```console
$ crio --version
crio version 1.24.3
Version:          1.24.3
GitCommit:        0e724229f2a81e08a5e05cef0b125ee0dadda08e
```

kubernetes is compile by ourself, which is forked from *https://github.com/kubernetes/kubernetes/tree/v1.25.8*

</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
# paste output here
$ uname -a
# paste output here
```

</details>

### Additional environment details (AWS, VirtualBox, physical, etc.)

<details>
VirtualBox
</details>

### PR DESCRIPTION
**Pull request by saschagrunert**
#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:
This new field allows fixing the kubelet image garbage collection in container runtimes. The `image_ref` has been historically used by container runtimes to reference images by digest.

#### Which issue(s) this PR fixes:

Allows to fix https://github.com/cri-o/cri-o/issues/7143

#### Special notes for your reviewer:
cc @mrunalp @haircommander @kubernetes/sig-node-api-reviews 
#### Does this PR introduce a user-facing change?

```release-note
Allowing container runtimes to fix an image garbage collection bug by adding an `image_id` field to the CRI Container message.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
None
```

### CODE REVIEW

**Comment by saschagrunert** [comment]
> thanks for picking this up! I believe for backwards compatibility with the kube pod API we need to change https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kubelet_pods.go#L1903 to be `ImageID: ImageRef` (if imageRef is != "", otherwise ImageID) or else we'll do the same thing as if CRI reported ImageID as ImageRef (new, broken cri-o behavior)

@haircommander the `ContainerStatus` has not changed in this PR, means we have no field `ImageRef` at all. We only changed the `Container` message, which is the one used for the image GC.

**Comment by sftim** [comment]
Does this PR make any change that a Kubernetes cluster operator could observe (even indirectly, but without using a debugger or trace logging)?

**Comment by saschagrunert** [comment]
> Does this PR make any change that a Kubernetes cluster operator could observe (even indirectly, but without using a debugger or trace logging)?

No, runtimes do not support that field yet so there will be no change. Only if runtimes set the value of the field, then it will influence how the image GC works.

**Comment by sftim** [comment]
>  runtimes do not support that field yet so there will be no change

If I used the release of Kubernetes that included this change, and a container runtime with support, would I notice the behavior? If so, we should change log it so people know it's available.

File: staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto [hunk 1]
```diff
@@ -1216,6 +1216,13 @@ message Container {
     // MUST be identical to that of the corresponding ContainerConfig used to
     // instantiate this Container.
     map<string, string> annotations = 9;
+    // Reference to the unique identifier of the image. The image_ref above has
```
Comment (mikebrow) [inline staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto:1219]: suggest the "For most runtimes.." sentence above in the description of image_ref

and for image_id suggest:

     // Reference to the unique identifier of the image, on the node, as returned in the image service apis.

**Review by mikebrow — COMMENTED** [review]
see suggested comment doc edits

**Comment by haircommander** [comment]
can someone help find where we propegate kuberuntime.ContainerStatus  to the kube API? From my perspective, we are translating kuberuntime.ImageID -> ContainerStatus.ImageID **somewhere** but I am struggling to find where. If that's the case, then it feels like this PR would change it so that we're reporting ImageID (which is coming from the CRI as an ImageID, not ImageRef) and then we'd change the ContainerStatus to have the hash and not repo + digest. It sounds like this isn't what is happening though

**Comment by saschagrunert** [comment]
> > runtimes do not support that field yet so there will be no change
> 
> If I used the release of Kubernetes that included this change, and a container runtime with support, would I notice the behavior? If so, we should change log it so people know it's available.

Yes, changed the release note for that.

File: staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto (hunk 1 above)
Comment (saschagrunert) [reply staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto:1219]: Changed as suggested.

**Comment by haircommander** [comment]
/lgtm

I was mixing up the kuberuntime.ContainerStatus with kuberuntime.Container. sorry for the noise!

**Comment by mikebrow** [comment]
> > fair to say we should also fix the status response in a similar fashion.. perhaps in another pr?
> 
> I'm wondering what we should fix there. We translate the container status to the kube API when we set the `kubecontainer.ImageID` from `status.ImageRef` (which is expected to be a digested reference right now):

Was thinking the image id might be useful. By fix I meant .. add/include in container status response.

**Comment by saschagrunert** [comment]
> add/include the image id in the container status response to go along with the image ref...

Yeah, I don't know. It would not be used in the first place and I personally would prefer developers relying on the field in the `Container` rather than the `ContainerStatus`. :thinking:

**Review by mikebrow — APPROVED** [review]
LGTM

**Comment by mikebrow** [comment]
> > add/include the image id in the container status response to go along with the image ref...
> 
> Yeah, I don't know. It would not be used in the first place and I personally would prefer developers relying on the field in the `Container` rather than the `ContainerStatus`. 🤔

just thinking they should be consistent... or at least both have both fields... It would help if crictl showed the imagery(ID) without requiring json output:
```
(base) root@ubnt:/home/mike/crictl-test# crictl ps -v -o json
{
  "containers": [
    {
      "id": "82ea9fe195bb46e4951a5f52b96e31e03dd09c1ba65c3c89ba2b2e7f0ce687c6",
      "podSandboxId": "e35c56275a8925ea0304ca11fbc82a2b0f8e6e581d6eea443b4a7a5784e216b9",
      "metadata": {
        "name": "busybox",
        "attempt": 0
      },
      "image": {
        "image": "busybox:1.35.0",
        "annotations": {
        },
        "userSpecifiedImage": "busybox:1.35.0",
        "runtimeHandler": ""
      },
      "imageRef": "sha256:0c00acac9c2794adfa8bb7b13ef38504300b505a043bf68dff7a00068dcc732b",
      "state": "CONTAINER_RUNNING",
      "createdAt": "1709135149290428214",
      "labels": {
      },
      "annotations": {
      }
    }
  ]
}
(base) root@ubnt:/home/mike/crictl-test# crictl -v
crictl version 1.29.0
(base) root@ubnt:/home/mike/crictl-test# crictl ps -v 
ID: 82ea9fe195bb46e4951a5f52b96e31e03dd09c1ba65c3c89ba2b2e7f0ce687c6
PodID: e35c56275a8925ea0304ca11fbc82a2b0f8e6e581d6eea443b4a7a5784e216b9
Name: busybox
Attempt: 0
State: Running
Image: busybox:1.35.0
Created: 37 minutes ago

(base) root@ubnt:/home/mike/crictl-test# crictl ps
CONTAINER           IMAGE               CREATED             STATE               NAME                ATTEMPT             POD ID              POD
82ea9fe195bb4       busybox:1.35.0      37 minutes ago      Running             busybox             0                   e35c56275a892       unknown

```

**Comment by mrunalp** [comment]
> Was thinking the image id might be useful. By fix I meant .. add/include the image id in the container status response to go along with the image ref...

I think we should avoid adding this to the pod status API unless there is a clear need for it. 
However, we should follow up to solidify the current pod status API for the containerStatus.imageID returning the ref alongwith tests so we don't regress on it.

**Review by mrunalp — APPROVED** [review]

**Comment by saschagrunert** [comment]
There is a image ID test available: https://github.com/kubernetes/kubernetes/blob/master/test/e2e_node/image_id_test.go

CRI-O still returns the image ID if the image is already present on disk, that's why the test allows both: https://github.com/kubernetes/kubernetes/blob/411c29c39f03687a30a8667295c61590def8fc89/test/e2e_node/image_id_test.go#L67-L68

**Comment by saschagrunert** [comment]
But there is a corner case when developers use `ConvertPodStatusToRunningPod` and override the field. Closing this gap in https://github.com/kubernetes/kubernetes/pull/123583

**Pull request merged**